File: src/main/java/com/boxfox/core/router/wallet/RemittanceRouter.java
package com.boxfox.core.router.wallet;

import com.boxfox.vertx.router.*;
import com.boxfox.vertx.service.*;
import com.boxfox.cross.service.WalletDatabaseService;
import com.boxfox.cross.wallet.WalletService;
import com.boxfox.cross.wallet.WalletServiceManager;
import com.boxfox.cross.wallet.model.TransactionResult;
import com.linkbit.android.entity.WalletModel;
import io.netty.handler.codec.http.HttpResponseStatus;
import io.vertx.core.http.HttpMethod;
import io.vertx.ext.web.RoutingContext;

public class RemittanceRouter extends AbstractRouter {

    @Service
    private WalletDatabaseService walletDatabaseService;

    @RouteRegistration(uri = "/remittance", method = HttpMethod.POST, auth = true)
    public void send(RoutingContext ctx,
                     @Param(name = "symbol") String symbol,
                     @Param(name = "walletName") String walletName,
                     @Param(name = "walletData") String walletData,
                     @Param(name = "password") String password,
                     @Param(name = "targetAddress") String targetAddress,
                     @Param(name = "amount") String amount) {
        WalletService service = WalletServiceManager.getService(symbol);
        walletDatabaseService.findByAddress(targetAddress, res -> {
            WalletModel wallet = res.result();
            if (wallet != null) {
                String destAddress = targetAddress;
                if (res != null)
                    destAddress = wallet.getAccountAddress();
                TransactionResult result = service.send(walletName, walletData, password, destAddress, amount);
                ctx.response().putHeader("content-type", "application/json");
                ctx.response().setChunked(true).write(gson.toJson(result));
            } else {
                ctx.response().setStatusCode(HttpResponseStatus.NOT_FOUND.code());
            }
            ctx.response().end();
        });
    }
}

File: src/main/java/com/boxfox/core/router/wallet/RemittanceRouter.kt
package com.boxfox.core.router.wallet

import com.boxfox.vertx.router.*
import com.boxfox.vertx.service.*
import com.boxfox.cross.service.wallet.WalletDatabaseService
import com.boxfox.linkbit.wallet.WalletServiceRegistry
import io.vertx.core.http.HttpMethod
import io.vertx.ext.web.RoutingContext

class RemittanceRouter : AbstractRouter() {

    @Service
    private lateinit var walletDatabaseService: WalletDatabaseService

    @RouteRegistration(uri = "/remittance", method = arrayOf(HttpMethod.POST))
    fun send(ctx: RoutingContext,
             @Param(name = "symbol") symbol: String,
             @Param(name = "walletName") walletName: String,
             @Param(name = "walletData") walletData: String,
             @Param(name = "password") password: String,
             @Param(name = "targetAddress") targetAddress: String,
             @Param(name = "amount") amount: String) {
        val service = WalletServiceRegistry.getService(symbol)
        walletDatabaseService.findByAddress(targetAddress).subscribe({
            var destAddress = it.accountAddress
            val result = service.send(walletName, walletData, password, destAddress, amount)
            ctx.response().putHeader("content-type", "application/json")
            ctx.response().setChunked(true).write(gson.toJson(result))
            ctx.response().end()
        }, {
            ctx.fail(it)
        })
    }
}
